src/main/java/de/dhbw/karlsruhe/top/down/parsing/validation/TopDownAcceptorValidation.java

```diff
@@ -0,0 +1,199 @@
+                success = this.validateReadStep(currStep, nextStep);
+            } else if (Arrays.stream(this.grammarService.getNonTerminals()).anyMatch(firstStackChar::equals)) {
+                success = this.validateProductionStep(currStep, nextStep);
+            }
+            if (!success) {
+                return success;
+            }
+        }
+
+        return success;
+    }
+
+    private boolean validateReadStep(TopDownStep tdStep, TopDownStep nextStep) {
+        boolean valid = false;
+
+        String stackNextStep = nextStep.getNewStack();
+        String stackCurrStep = tdStep.getNewStack();
+        String inputNextStep = nextStep.getReadInput();
+        GrammarRule productionCurrStep = tdStep.getUsedProduction();
+        TopDownState stateCurrStep = tdStep.getNewState();
+
+        if (stackCurrStep == null || stateCurrStep != TopDownState.Z) {
+            return false;
+        }
+
+        if (this.checkReadInput(stackCurrStep, inputNextStep) && this.checkSideConditionsRead(stackCurrStep, stackNextStep, productionCurrStep)) {
+            valid = true;
+        }
+
+        return valid;
+    }
+
+    private boolean checkReadInput(String stackCurrStep, String inputNextStep) {
+        boolean success = false;
+
+        char firstCharStackCurrStep = stackCurrStep.charAt(0);
+        char firstCharInputNextStep = inputNextStep.charAt(inputNextStep.length()-1);
+
+        if (firstCharStackCurrStep == firstCharInputNextStep) {
+            success = true;
+        }
+
+        return success;
+    }
+
+    private boolean checkSideConditionsRead(String stackCurrStep, String stackNextStep, GrammarRule productionCurrStep) {
+        boolean success = false;
+
+        char secondCharStackCurrStep = stackCurrStep.charAt(1);
+        char firstCharStackNextStep = stackNextStep.charAt(0);
+
+        if (productionCurrStep == null && (secondCharStackCurrStep == firstCharStackNextStep)) {
+            success = true;
+        }
+
+        return success;
+    }
+
+    private boolean validateProductionStep(TopDownStep tdStep, TopDownStep nextStep) {
+        boolean valid = false;
+
+        String stackCurrStep = tdStep.getNewStack();
+        String stackNextStep = nextStep.getNewStack();
+        GrammarRule productionCurrStep = tdStep.getUsedProduction();
+        String inputCurrStep = tdStep.getReadInput();
+        String inputNextStep = nextStep.getReadInput();
+        TopDownState stateCurrStep = tdStep.getNewState();
+
+        if (stackCurrStep == null || productionCurrStep == null || stateCurrStep != TopDownState.Z) {
+            return false;
+        }
+
+        if (this.checkUsedProduction(stackCurrStep, stackNextStep, productionCurrStep) && this.checkSideConditionsProduction(inputCurrStep, inputNextStep)) {
+            valid = true;
+        }
+
+        return valid;
+    }
+
+    private boolean checkUsedProduction(String stackCurrStep, String stackNextStep, GrammarRule productionCurrStep) {
```

Comment:
Habe gerade keine Stelle gefunden an der geprüft wird, ob die Produktion in der Grammatik vorhanden ist. Wenn noch nicht implementiert, vielleicht in dieser Methode hinzufügen

Reply:
So wie ich das sehe, wird mit überprüft, ob die Produktion in der Grammatik enthalten ist. Denn wenn es die Produktion gar nicht geben würde, dann würde die Linke Seite nicht mit der richtigen linken Seite und die rechte Seite nicht mit der korrekten rechten Seite einer Produktion übereinstimmen. Diese Validierung wird in checkUsedProduction validiert oder übersehe ich da was bei der Implementierung?